### PR TITLE
Dials

### PR DESCRIPTION
**Pull request by jyoo84**
Basic prototype of a maneuver dial  that loads a test dial datum  and renders a dial which displays the selected maneuver.

### CODE REVIEW

File: src/Dials.js [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+            xws: xwsm,
+            speed: parseInt(xwsm[0], 10),
+            bearing: xwsm[1],
+            // May be useful to convert this to a more semantic value, but keeping the xws difficulty for now
+            difficulty: xwsm[2]
+        }
+    })
+}
+
+// to be used with var unique = a.filter(onlyUnique) where a is an array
```
Comment (jcavanagh) [inline src/Dials.js]: Let's remove unused comments

File: src/Dials.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+
+// function MButton(props) {
+//     return (
+        
+//     );
+// }
+
+class MDial extends React.Component {
+    // put in some state for now
```
Comment (jcavanagh) [inline src/Dials.js]: Unused comments

File: src/Dials.js [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    //     </>
+    //     );
+    // }
+
+    renderButton(maneuverData) {
+        const buttonSize = '35px';
+        const border = '1px solid';
+        const display = 'flex';
+        // added so I could debug
+        var mymdata = maneuverData;
```
Comment (jcavanagh) [inline src/Dials.js]: Old debugging code

File: src/Dials.js [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        }
+
+        // Placeholder button
+        return <div style={{ height: buttonSize, width: buttonSize, border, display }} />;
+    }
+
+    render() {
+        const parsedManeuvers = parseXwsManeuvers(this.props.maneuvers);
+        const bySpeed = groupBy(parsedManeuvers, 'speed');
+        // const existingBearings = getMBearings(this.props.maneuvers);
```
Comment (jcavanagh) [inline src/Dials.js]: Unused comments

File: src/index.js [hunk 5]
```diff
@@ -13,7 +13,8 @@ import 'firebase/database';
 import 'firebase/firestore';
 import 'firebase/functions';
 
-import App from './App';
+// import App from './App';
```
Comment (jcavanagh) [inline src/index.js]: This should be reset to the normal app import

File: src/index.js [hunk 6]
```diff
@@ -26,4 +27,5 @@ firebase.initializeApp({
   measurementId: 'G-WH9BPRDX4K',
 });
 
-ReactDOM.render(<App />, document.getElementById('root'));
+// ReactDOM.render(<App />, document.getElementById('root'));
```
Comment (jcavanagh) [inline src/index.js]: This should be reset to the normal app entry point - we can make a separate shim for independent dials testing

File: src/Dials.js [hunk 7]
```diff
@@ -0,0 +1,367 @@
+import React from 'react';
+
+import groupBy from 'lodash/groupBy';
+import mapValues from 'lodash/mapValues';
+import range from 'lodash/range';
+
+// this is a maneuver button, e.g., 1 straight, 2  turn, 3 bank, 4-k etc..
+const test_dial = [
```
Comment (JeremyHurwitz) [inline src/Dials.js]: Keep test code out of the main code. This stuff goes in Dials.test.js.

File: src/Dials.js [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+    "3FB",
+    "3NW",
+    "3YW",
+    "3RR",
+    "4FW",
+    "4KR",
+    "5KR"
+]
+
+// order of columns of bearings which  should appear if they  exist.
```
Comment (JeremyHurwitz) [inline src/Dials.js]: typos (double space)

File: src/Dials.js (hunk 7 above)
Comment (JeremyHurwitz) [inline src/Dials.js]: Keep test code out of the main code. This stuff goes in Dials.test.js.

File: src/Dials.js [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+const bearingOrder = [
+    "T",
```
Comment (JeremyHurwitz) [inline src/Dials.js]: What do these letters mean?

File: src/Dials.js [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    "E",
+    "R",
+    "O",
+    "A",
+    "P",
+    "D"
+]
+
+// https://github.com/raithos/xwing/blob/master/coffeescripts/xwing.coffee#L1854
+const bearingSvg = {
```
Comment (JeremyHurwitz) [inline src/Dials.js]: Personal preference, but I would probably hide this off in its own file.

Comment (JeremyHurwitz) [reply src/Dials.js]: Maybe just export renderManeuverSvg() from that file.

File: src/Dials.js [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+        triangle: 'M60,100 H100 L80,140 Z',
+        transform: null
+    },
+    "R": {
+        line: 'M40,180 L40,70 120,70',
+        triangle: 'M100,100 H140 L120,140 Z',
+        transform: null
+    },
+    "O": {
+        element: (maneuver, color) => {
```
Comment (JeremyHurwitz) [inline src/Dials.js]: Why is this one different from all the others? What is 'element'?

File: src/Dials.js (hunk 1 above)
Comment (JeremyHurwitz) [inline src/Dials.js]: Delete?

Comment (JeremyHurwitz) [reply src/Dials.js]: Ditto throughout for the commented out code.

File: src/Dials.js [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+            </g>
+        </svg>
+    )
+}
+
+function parseXwsManeuvers(xwsManeuvers) {
+    return xwsManeuvers.map(xwsm => {
+        return {
+            xws: xwsm,
+            speed: parseInt(xwsm[0], 10),
```
Comment (JeremyHurwitz) [inline src/Dials.js]: How does this handle the quadjumper? That has speed = -1.

File: src/Dials.js (hunk 9 above)
Comment (jcavanagh) [reply src/Dials.js]: These are xws-spec magic values.  They are not particularly documented, but represent the possible maneuver bearings e.g: https://github.com/guidokessels/xwing-data2/blob/master/data/pilots/first-order/tie-ba-interceptor.json#L6

File: src/Dials.js (hunk 11 above)
Comment (jcavanagh) [reply src/Dials.js]: This is the only one for which the svg element is a `rect` and not a `path`.  `element` is an optional property whose value is a function that returns an SVG element.  

Used as an alternative to the default implementation of drawing 2 paths with maneuver line + triangle path data.

Comment (jcavanagh) [reply src/Dials.js]: Also I lifted these straight from YASB, and could have made it a path but I was too lazy

File: src/Dials.js (hunk 12 above)
Comment (jcavanagh) [reply src/Dials.js]: The data encodes reverse as a bearing, not as a negative speed:  https://github.com/guidokessels/xwing-data2/blob/master/data/pilots/scum-and-villainy/quadrijet-transfer-spacetug.json#L7

File: src/Dials.js (hunk 9 above)
Comment (JeremyHurwitz) [reply src/Dials.js]: If we're stuck with the magic values, then comments would be great
  "T",  // straight
  "B",  // bank left
  "F",  // hard left
etc.

File: src/index.js (hunk 6 above)
Comment (JeremyHurwitz) [reply src/index.js]: It should be possible to import & test Dials.js directly.

**Pull request merged**